### ISSUE TITLE
Issue with backtesting multiple instruments using data catalog

### ISSUE THREAD

**Issue by amitsinghpgs** [nautechsystems/nautilus_trader#560]
### Expected Behavior
When multiple instruments are present in a backtest then the data with same time stamp should be sent to backtest engine at the same time.

### Actual Behavior
I see that data for instrument A was sent for 2019- 2020 then  then data for instrument B was sent for 2019. This does not mimic the real world behaviour.

Data sample that I am using:
```
price           size aggressor_side  trade_id             ts_event              ts_init
0  4.42400000    40.20000000            BUY  42411134  1641229318055000064  1641229318055000064
1  4.42400000    27.20000000            BUY  42411135  1641229318055000064  1641229318055000064
2  4.42400000  1567.60000000            BUY  42411136  1641229318055000064  1641229318055000064
3  4.42500000    22.60000000           SELL  42411138  1641229318191000064  1641229318191000064
4  4.42500000    22.60000000           SELL  42411137  1641229318191000064  1641229318191000064
```

Parquet file folder:
```
$ ls -l '/home/amit/catalog/data/trade_tick.parquet/instrument_id=KAVA-USDT.BINANCE'
total 774756
-rw-rw-r-- 1 amit amit 70033279 Feb  2 08:43 1571968800000000000-1596897580134000128-0.parquet
-rw-rw-r-- 1 amit amit 64047472 Feb  2 08:44 1596897580134000128-1605822125964999936-0.parquet
-rw-rw-r-- 1 amit amit 64811784 Feb  2 08:45 1605822127964999936-1612926729784000000-0.parquet
-rw-rw-r-- 1 amit amit 62435778 Feb  2 08:46 1612926729799000064-1615602833971000064-0.parquet
-rw-rw-r-- 1 amit amit 61395462 Feb  2 08:47 1615602835135000064-1618142491108999936-0.parquet
-rw-rw-r-- 1 amit amit 61060029 Feb  2 08:48 1618142491118000128-1620721070134000128-0.parquet
-rw-rw-r-- 1 amit amit 63533327 Feb  2 08:49 1620721070348999936-1623103529528000000-0.parquet
-rw-rw-r-- 1 amit amit 64664459 Feb  2 08:50 1623103529528000000-1627355135007000064-0.parquet
-rw-rw-r-- 1 amit amit 61964848 Feb  2 08:51 1627355135852999936-1629554430359000064-0.parquet
-rw-rw-r-- 1 amit amit 59786230 Feb  2 08:52 1629554430359000064-1631892647757999872-0.parquet
-rw-rw-r-- 1 amit amit 61911957 Feb  2 08:53 1631892647757999872-1635272956584999936-0.parquet
-rw-rw-r-- 1 amit amit 63083945 Feb  2 08:54 1635272958720999936-1639708285194000128-0.parquet
-rw-rw-r-- 1 amit amit 34581881 Feb  2 08:54 1639708285324000000-1642068403552999936-0.parquet
```

backtest logs showing the issue
```
[1m2020-11-26T06:27:37.003000064Z[0m [INF] BACKTESTER-000.TurtleTrade_1_3-000: Received Bar Bar(YFI/USDT.BINANCE-250-TICK-LAST-INTERNAL,21078.70000000,21150.00000000,21054.43000000,21150.00000000,12.36640700,1606372057003000064)[0m
[1m2020-11-26T06:28:14.324999936Z[0m [INF] BACKTESTER-000.TurtleTrade_1_3-000: Received Bar Bar(YFI/USDT.BINANCE-250-TICK-LAST-INTERNAL,21150.00000000,21200.00000000,21108.98000000,21134.33000000,5.20373000,1606372094324999936)[0m
[1m2020-11-26T06:30:00.674000128Z[0m [INF] BACKTESTER-000.TurtleTrade_1_3-000: Received Bar Bar(YFI/USDT.BINANCE-250-TICK-LAST-INTERNAL,21134.33000000,21200.00000000,21100.32000000,21200.00000000,9.41785400,1606372200674000128)[0m
[1m2020-11-26T06:30:54.672000000Z[0m [INF] BACKTESTER-000.TurtleTrade_1_3-000: Received Bar Bar(YFI/USDT.BINANCE-250-TICK-LAST-INTERNAL,21200.00000000,21329.25000000,21200.00000000,21299.69000000,7.46553100,1606372254672000000)[0m
[1m2020-11-26T06:32:04.936999936Z[0m [INF] BACKTESTER-000.TurtleTrade_1_3-000: Received Bar Bar(YFI/USDT.BINANCE-250-TICK-LAST-INTERNAL,21299.69000000,21345.00000000,21237.20000000,21309.05000000,9.53524400,1606372324936999936)[0m
[1m2020-11-26T06:33:05.604000000Z[0m [INF] BACKTESTER-000.TurtleTrade_1_3-000: Received Bar Bar(YFI/USDT.BINANCE-250-TICK-LAST-INTERNAL,21309.05000000,21345.00000000,21199.00000000,21200.00000000,8.23452300,1606372385604000000)[0m
[1m2020-11-26T06:33:27.648999936Z[0m [INF] BACKTESTER-000.TurtleTrade_1_3-000: Received Bar Bar(YFI/USDT.BINANCE-250-TICK-LAST-INTERNAL,21200.00000000,21249.38000000,21100.00000000,21100.00000000,19.66040700,1606372407648999936)[0m
[1m2020-11-26T06:33:40.529999872Z[0m [INF] BACKTESTER-000.TurtleTrade_1_3-000: Received Bar Bar(YFI/USDT.BINANCE-250-TICK-LAST-INTERNAL,21100.00000000,21147.93000000,21049.93000000,21049.93000000,24.55945000,1606372420529999872)[0m
[1m2022-02-01T18:47:39.870220364Z[0m [INF] BACKTESTER-000.BacktestEngine: Added 5,212,102 YFI/USDT.BINANCE TradeTick elements.[0m
[1m2022-02-01T18:47:39.870550481Z[0m [INF] BACKTESTER-000.BacktestEngine: [36m=================================================================[0m
[1m2022-02-01T18:47:39.870572669Z[0m [INF] BACKTESTER-000.BacktestEngine: [36m BACKTEST RUN[0m
[1m2022-02-01T18:47:39.870588235Z[0m [INF] BACKTESTER-000.BacktestEngine: [36m=================================================================[0m
[1m2022-02-01T18:47:39.870602067Z[0m [INF] BACKTESTER-000.BacktestEngine: Run config ID:  a7c12c17b28f78ac66255d75fc42a2bd[0m
[1m2022-02-01T18:47:39.870616891Z[0m [INF] BACKTESTER-000.BacktestEngine: Run ID:         a05c585d-bc78-4c11-9d19-f6b64a46627c[0m
[1m2022-02-01T18:47:39.870635434Z[0m [INF] BACKTESTER-000.BacktestEngine: Run started:    2022-02-01 18:38:57.919154+00:00[0m
[1m2022-02-01T18:47:39.870651435Z[0m [INF] BACKTESTER-000.BacktestEngine: Backtest start: 2019-10-25 02:00:00+00:00[0m
[1m2022-02-01T18:47:39.870667208Z[0m [INF] BACKTESTER-000.BacktestEngine: Batch start:    2019-09-29 04:00:00+00:00.[0m
[1m2022-02-01T18:47:39.870683376Z[0m [INF] BACKTESTER-000.BacktestEngine: Batch end:      2021-01-02 19:01:27.697999872+00:00.[0m
[1m2022-02-01T18:47:39.870695110Z[0m [INF] BACKTESTER-000.BacktestEngine: [36m-----------------------------------------------------------------[0m
[1m2019-09-29T04:09:13.323000064Z[0m [INF] BACKTESTER-000.TurtleTrade_1_3-000: Received Bar Bar(HBAR/USDT.BINANCE-250-TICK-LAST-INTERNAL,0.03103000,0.04147000,0.03103000,0.04109000,1603779.60000000,1569730153323000064)[0m
[1m2019-09-29T04:09:13.323000064Z[0m [94m[INF] BACKTESTER-000.TurtleTrade_1_3-000: Waiting for indicators to warm up [1]...[0m
[1m2019-09-29T04:17:45.493999872Z[0m [INF] BACKTESTER-000.TurtleTrade_1_3-000: Received Bar Bar(HBAR/USDT.BINANCE-250-TICK-LAST-INTERNAL,0.04109000,0.04400000,0.03937000,0.04076000,1396585.00000000,1569730665493999872)[0m
[1m2019-09-29T04:17:45.493999872Z[0m [94m[INF] BACKTESTER-000.TurtleTrade_1_3-000: Waiting for indicators to warm up [2]...[0m
[1m2019-09-29T04:45:31.812999936Z[0m [INF] BACKTESTER-000.TurtleTrade_1_3-000: Received Bar Bar(HBAR/USDT.BINANCE-250-TICK-LAST-INTERNAL,0.04076000,0.04076000,0.03866000,0.04015000,1273388.90000000,1569732331812999936)[0m
[1m2019-09-29T04:45:31.812999936Z[0m [94m[INF] BACKTESTER-000.TurtleTrade_1_3-000: Waiting for indicators to warm up [3]...[0m
```

**Comment by amitsinghpgs** [comment]
@limx0 I am attaching a sample code which you can directly plug in. The following code has the catalog creation step and also it uses `batch_files` function to read the batches of data. As you will see in the output at the end the batches are not chronological



```

files = {
    "KAVA": "/home/amit/DATA/crypto_archive/KAVAUSDT.csv.gz",
}
instruments = {}
for curr in files:
    instruments[curr] = CurrencySpot(
        instrument_id=InstrumentId(
            symbol=Symbol(f"{curr}/USDT"),
            venue=Venue("BINANCE"),
        ),
        native_symbol=Symbol(f"{curr}USDT"),
        base_currency=Currency.from_str(curr),
        quote_currency=USDT,
        price_precision=8,
        size_precision=8,
        price_increment=Price(1e-08, precision=8),
        size_increment=Quantity(1e-08, precision=8),
        lot_size=None,
        max_quantity=Quantity(1e10, precision=8),
        min_quantity=Quantity(1e-08, precision=8),
        max_notional=None,
        min_notional=Money(1e-08, USDT),
        max_price=Price(1e10, precision=8),
        min_price=Price(1e-08, precision=8),
        margin_init=Decimal("1.00"),
        margin_maint=Decimal("0.35"),
        maker_fee=Decimal("0.0001"),
        taker_fee=Decimal("0.0001"),
        ts_event=0,
        ts_init=0,
    )
instruments


import os, shutil

CATALOG_PATH =  "/home/amit//catalog"
# Clear if it already exists, then create fresh
if os.path.exists(CATALOG_PATH):
    shutil.rmtree(CATALOG_PATH)
os.mkdir(CATALOG_PATH)




CATALOG_PATH =  "/home/amit//catalog"
catalog = DataCatalog(CATALOG_PATH)


write_objects(catalog,list(instruments.values()))



def parser(data, instrument):
    if data is None:
        return
    logging.info(f"{repr(instrument.id)} data chunk size {len(data)}")
    print(data.head())
    data['side'] = data['side'].astype(int).apply(lambda x: "BUY" if x==1 else "SELL")
    data.loc[:, "timestamp"] = pd.to_datetime(data["timestamp"], unit='ms')
    wrangler = TradeTickDataWrangler(instrument)
    ticks = wrangler.process(data.set_index("timestamp"))
    yield from ticks



for curr in files:
    process_files(
        glob_path=files[curr],
        reader=CSVReader(
            block_parser=lambda x: parser(x, instrument=instruments[curr]), 
            header=['trade_id', 'price', 'quantity',  'timestamp', 'side'],
            chunked=True, 
            as_dataframe=True,
            separator='|'
        ),
        catalog=catalog,
    )


# # Read data from the catalog and print out min and max date in each batch
# As can be seen the min and max date in each batch does not follow a chronological order



CATALOG_PATH = "/home/amit/catalog"
catalog = DataCatalog(CATALOG_PATH)
start = dt_to_unix_nanos(pd.Timestamp("2019-01-01", tz="UTC"))
end = dt_to_unix_nanos(pd.Timestamp("2022-10-30", tz="UTC"))
data_config = [
    BacktestDataConfig(
        catalog_path=CATALOG_PATH,
        data_cls_path="nautilus_trader.model.data.tick.TradeTick",
        instrument_id=instrument.id.value,
        start_time=start,
        end_time=end,
    )
    for instrument in catalog.instruments(as_nautilus=True)
]
for f in batch_files(
    catalog,
    data_config,
):
    dates = [dt.datetime.fromtimestamp(x.ts_init / 1e9) for x in f]
    print(f"min date : {min(dates)}, max date: {max(dates)}")
```



The output i see is mentioned below. min, max date show that batches are not chronological
```
min date : 2020-07-07 21:35:06.790000, max date: 2020-08-08 09:39:40.134000
min date : 2019-10-24 21:00:00, max date: 2020-08-02 10:00:59.714000
min date : 2020-01-24 17:00:29.331000, max date: 2020-06-16 04:07:26.323000
min date : 2020-06-16 04:07:26.323000, max date: 2020-11-19 15:42:05.965000
min date : 2020-09-04 08:32:06.881000, max date: 2020-10-30 01:43:25.712000
min date : 2020-08-08 09:39:40.134000, max date: 2020-11-10 12:53:14.863000
min date : 2020-08-12 14:20:35.155000, max date: 2021-02-09 21:12:09.784000
min date : 2020-11-22 12:04:08.495000, max date: 2021-01-06 08:35:54.761000
min date : 2021-01-06 08:35:54.762000, max date: 2021-01-23 06:51:11.495000
min date : 2021-01-23 06:51:11.495000, max date: 2021-03-12 09:31:13.082000
min date : 2021-02-09 21:12:09.799000, max date: 2021-03-12 20:33:53.971000
min date : 2021-02-17 04:08:51.627000, max date: 2021-02-28 21:11:07.694000
min date : 2021-02-28 21:11:07.821000, max date: 2021-03-11 23:15:17.814000
min date : 2021-03-11 23:15:17.814000, max date: 2021-04-11 07:01:31.109000
min date : 2021-03-12 20:33:55.135000, max date: 2021-03-30 13:15:18.710000
min date : 2021-03-16 16:43:14.518000, max date: 2021-04-05 06:41:38.253000
min date : 2021-04-05 06:41:38.294000, max date: 2021-05-11 03:17:50.134000
min date : 2021-04-11 07:01:31.118000, max date: 2021-05-02 00:32:26.593000
min date : 2021-04-16 09:25:19.143000, max date: 2021-05-04 21:20:04.934000
min date : 2021-05-04 21:20:04.996000, max date: 2021-06-07 17:05:29.528000
min date : 2021-05-22 23:38:26.028000, max date: 2021-06-02 08:38:59.499000
min date : 2021-05-11 03:17:50.349000, max date: 2021-06-05 00:10:46.910000
min date : 2021-05-12 19:02:13.957000, max date: 2021-07-26 19:00:42.431000
min date : 2021-07-07 16:04:43.241000, max date: 2021-07-26 22:05:35.007000
min date : 2021-06-22 07:54:10.829000, max date: 2021-07-26 04:29:15.025000
min date : 2021-06-07 17:05:29.528000, max date: 2021-07-07 16:04:43.226000
min date : 2021-06-22 00:23:37.667000, max date: 2021-08-21 09:00:30.359000
min date : 2021-08-02 09:33:34.864000, max date: 2021-08-18 02:20:59.886000
min date : 2021-08-03 06:09:15.358000, max date: 2021-08-21 00:58:14.761000
min date : 2021-08-11 01:19:40.096000, max date: 2021-09-17 10:30:47.758000
min date : 2021-08-22 09:14:01.126000, max date: 2021-08-28 18:20:24.679000
min date : 2021-08-28 18:20:26.547000, max date: 2021-09-05 09:38:37.460000
min date : 2021-09-05 09:38:37.460000, max date: 2021-10-26 03:05:31.737000
min date : 2021-09-25 20:00:23.432000, max date: 2021-10-26 13:29:16.585000
min date : 2021-09-17 10:30:47.758000, max date: 2021-10-06 08:33:59.232000
min date : 2021-09-25 07:15:46.796000, max date: 2021-10-20 03:25:36.026000
min date : 2021-10-20 03:25:36.026000, max date: 2021-12-16 20:31:25.194000
min date : 2021-10-31 12:01:32.931000, max date: 2021-11-13 08:14:53.846000
min date : 2021-11-13 08:14:53.846000, max date: 2021-11-30 04:48:09.254000
min date : 2021-11-30 04:48:10.128000, max date: 2022-01-08 15:42:17.364000
min date : 2021-12-16 20:31:25.324000, max date: 2022-01-13 04:06:43.553000
min date : 2021-12-29 10:06:01.658000, max date: 2022-01-03 14:02:23.053000
```

**Comment by limx0** [comment]
Thanks for the example code @amitsinghpgs. I'm actually having trouble replicating this. Heres what I've done: 

- Copied your sample code
- Downloaded some sample data (daily bars) from https://www.cryptodatadownload.com/data/binance/
- I split the single file into a file for each year (2019-2022) and renamed them to be purposefully confusing:
  - 01_2021.csv
  - 02_2019.csv
  - 03_2020.csv 
  - 04_2022.csv
- Ran your code with some slight modifications (removed duplicate catalog path references and pointed to my raw data)
- Reduced the `target_batch_size` to `10kb` to force batches of data

When I do this, everything works as I would expect:

The catalog has the following saved files: 

```
1547596800000000000-1577750400000000000-0.parquet
1577836800000000000-1609372800000000000-0.parquet
1609459200000000000-1640908800000000000-0.parquet
1640995200000000000-1644105600000000000-0.parquet
```

Your final code block prints dates in order:
```
min date : 2019-01-16 11:00:00, max date: 2019-12-31 11:00:00
min date : 2020-01-01 11:00:00, max date: 2020-12-31 11:00:00
min date : 2021-01-01 11:00:00, max date: 2021-12-31 11:00:00
min date : 2022-01-01 11:00:00, max date: 2022-02-06 11:00:00
```

Here's my modified version of your code: 

<details>

```python
import datetime as dt
import logging
import os
import shutil
from decimal import Decimal

import pandas as pd
from dask.utils import parse_bytes
from nautilus_trader.backtest.config import BacktestDataConfig
from nautilus_trader.backtest.data.wranglers import TradeTickDataWrangler
from nautilus_trader.core.datetime import dt_to_unix_nanos
from nautilus_trader.model.currencies import USDT
from nautilus_trader.model.currency import Currency
from nautilus_trader.model.identifiers import InstrumentId
from nautilus_trader.model.identifiers import Symbol
from nautilus_trader.model.identifiers import Venue
from nautilus_trader.model.instruments.currency import CurrencySpot
from nautilus_trader.model.objects import Money
from nautilus_trader.model.objects import Price
from nautilus_trader.model.objects import Quantity
from nautilus_trader.persistence.batching import batch_files
from nautilus_trader.persistence.catalog import DataCatalog
from nautilus_trader.persistence.external.core import process_files
from nautilus_trader.persistence.external.core import write_objects
from nautilus_trader.persistence.external.readers import CSVReader


CATALOG_PATH = f"{os.environ['HOME']}/Downloads/Cryptodata/catalog"

files = {
    "LINK": f"{os.environ['HOME']}/Downloads/Cryptodata/raw/*.csv"
}
instruments = {}
for curr in files:
    instruments[curr] = CurrencySpot(
        instrument_id=InstrumentId(
            symbol=Symbol(f"{curr}/USDT"),
            venue=Venue("BINANCE"),
        ),
        native_symbol=Symbol(f"{curr}USDT"),
        base_currency=Currency.from_str(curr),
        quote_currency=USDT,
        price_precision=8,
        size_precision=8,
        price_increment=Price(1e-08, precision=8),
        size_increment=Quantity(1e-08, precision=8),
        lot_size=None,
        max_quantity=Quantity(1e10, precision=8),
        min_quantity=Quantity(1e-08, precision=8),
        max_notional=None,
        min_notional=Money(1e-08, USDT),
        max_price=Price(1e10, precision=8),
        min_price=Price(1e-08, precision=8),
        margin_init=Decimal("1.00"),
        margin_maint=Decimal("0.35"),
        maker_fee=Decimal("0.0001"),
        taker_fee=Decimal("0.0001"),
        ts_event=0,
        ts_init=0,
    )


# Clear if it already exists, then create fresh
if os.path.exists(CATALOG_PATH):
    shutil.rmtree(CATALOG_PATH)
os.mkdir(CATALOG_PATH)

catalog = DataCatalog(CATALOG_PATH)


write_objects(catalog, list(instruments.values()))


def parser(data, instrument):
    if data is None:
        return
    logging.info(f"{repr(instrument.id)} data chunk size {len(data)}")
    print(data.head())
    # data["side"] = data["side"].astype(int).apply(lambda x: "BUY" if x == 1 else "SELL")
    data.loc[:, "timestamp"] = pd.to_datetime(data["unix"], unit="ms")
    data.loc[:, "price"] = data['close']
    data.loc[:, "quantity"] = data['Volume LINK']
    data.loc[:, "buyer_maker"] = False
    data.loc[:, "trade_id"] = data['date']
    wrangler = TradeTickDataWrangler(instrument)
    ticks = wrangler.process(data.set_index("timestamp"))
    yield from ticks


for curr in files:
    process_files(
        glob_path=files[curr],
        reader=CSVReader(
            block_parser=lambda x: parser(x, instrument=instruments[curr]),
            # header=["trade_id", "price", "quantity", "timestamp", "side"],
            chunked=True,
            as_dataframe=True,
            # separator="|",
        ),
        catalog=catalog,
    )


# # Read data from the catalog and print out min and max date in each batch
# As can be seen the min and max date in each batch does not follow a chronological order


start = dt_to_unix_nanos(pd.Timestamp("2019-01-01", tz="UTC"))
end = dt_to_unix_nanos(pd.Timestamp("2022-10-30", tz="UTC"))
data_config = [
    BacktestDataConfig(
        catalog_path=CATALOG_PATH,
        data_cls_path="nautilus_trader.model.data.tick.TradeTick",
        instrument_id=instrument.id.value,
        start_time=start,
        end_time=end,
    )
    for instrument in catalog.instruments(as_nautilus=True)
]
for f in batch_files(catalog, data_config, target_batch_size_bytes=parse_bytes("10kb")):
    dates = [dt.datetime.fromtimestamp(x.ts_init / 1e9) for x in f]
    print(f"min date : {min(dates)}, max date: {max(dates)}")

```
</details>

Here is the raw data and catalog that was generated: [Cryptodata.zip](https://github.com/nautechsystems/nautilus_trader/files/8011223/Cryptodata.zip)

If you see any issues with my setup above could you let me know? 

Otherwise, I think at this stage I'm going to need to get some sample data from you in order to debug this, would that be a possibility?

**Comment by amitsinghpgs** [comment]
@limx0 i checked your changes (https://github.com/nautechsystems/nautilus_trader/pull/568) with one instrument and so far it looks good. Before we close this issue I will go ahead and check the `batch_files` function against multiple instruments and I will run a full backtest with multiple instruments to see that data is coming chronologically in the backtest and not just in `batch_files` function..

**Comment by TheRealBecks** [comment]
It's still not working for me. I deleted my data and added it to the catalog once again. Afterwards I ran my backtest, but:
- Data from the past or future is added in between
- The log is corrupted
- There is also byte data in the log (the file is way too big, so no attachment)

Some example screenshot from the log:
![grafik](https://user-images.githubusercontent.com/1497405/153477878-e4d21785-9065-4b3b-bfa3-42c1ce282866.png)

The issue isn't exactly reproducible, but it happens multiple time in the backtest. The longer the backtest is running the more it's going to happen.

I attached my files ([backtest.zip](https://github.com/nautechsystems/nautilus_trader/files/8043606/backtest.zip)) and the data to this comment:
- Load the data to the catalog by running `nautilus3.py`
- Run the backtest by running `nautilus4.py`
- Wait till the start of 2020. It's likely that you will see the heavily changing timestamps already directly in the console

One question: In `nautilus3.py` I have the function `make_equity_instrument` with the argument `price_precision=2`, but I found out that some data entries have a precision of e.g. 4:
```
2018-12-24 08:06:00,16.55,16.55,16.55,16.55,231
2018-12-24 08:07:00,16.5401,16.55,16.5401,16.55,3349
```
=> `16.5401`

Could that lead to an issue here?

**Comment by amitsinghpgs** [comment]
@limx0 i tested out the batch_files function with a large dataset having multiple instruments and initial impression seems that data chunks are coming out in chronological order.

**Comment by amitsinghpgs** [comment]
@limx0 @cjdsellers i ran a full backtest with around 4 GB of data today and it seems now that issue is resolved.

The `batch_files` function is returning data in correct order.
The backtest is receiving data in correct order.

Regards

**Comment by cjdsellers** [comment]
Many thanks to @limx0 @amitsinghpgs and @TheRealBecks for your efforts on this!

**Comment by amitsinghpgs** [comment]
@limx0 https://github.com/nautechsystems/nautilus_trader/blob/master/nautilus_trader/persistence/batching.py#L52 runs into error when dataframe is empty. It might be helpful if this piece of code is tested more extensively. I think we need to change the code to below so as to check that `df `is not empty
```

if file_meta.instrument_id and not df.empty:
        df.loc[:, "instrument_id"] = file_meta.instrument_id
```

**Comment by cjdsellers** [comment]
@TheRealBecks are you still running into issues related to this? or did the fix work for you as well?

**Comment by cjdsellers** [comment]
> @limx0 https://github.com/nautechsystems/nautilus_trader/blob/master/nautilus_trader/persistence/batching.py#L52 runs into error when dataframe is empty. It might be helpful if this piece of code is tested more extensively. I think we need to change the code to below so as to check that `df `is not empty
> 
> ```
> 
> if file_meta.instrument_id and not df.empty:
>         df.loc[:, "instrument_id"] = file_meta.instrument_id
> ```

+1 for more testing. I have it near the top of my list to add more tests and then do a refactor and docs sweep

**Comment by TheRealBecks** [comment]
I tested with the current version 138 (stable, not develop, because compilation with gcc failed) and I still have the same issue:
```
2020-05-14T12:05:00.000000000Z [INF] BACKTESTER-001.EMACross-000: Received Bar(AMD.NYSE-1-MINUTE-LAST-EXTERNAL,52.5,52.62,52.49,52.61,86126,1589457900000000000)
2019-04-26T15:50:00.000000000Z [INF] BACKTESTER-001.Portfolio: AMD.NYSE net_position=-1_000
2020-05-14T12:06:00.000000000Z [INF] BACKTESTER-001.EMACross-000: Received Bar(AMD.NYSE-1-MINUTE-LAST-EXTERNAL,52.59,52.65,52.5,52.62,87271,1589457960000000000)
```

**Comment by cjdsellers** [comment]
Fixed as of `1.140.0`
